core/embed/rust/src/ui/display.rs

```diff
@@ -402,6 +410,745 @@ pub fn rect_rounded2_partial(
+                        || !is_clockwise_or_equal_inc(n_vx, end_vector)
+                } else {
+                    is_clockwise_or_equal(n_start, vx)
+                        && is_clockwise_or_equal_inc(n_vx, end_vector)
+                };
+
+                // The antialiasing calculation below uses simplified distance difference
+                // calculation. Optimally, SQRT should be used, but assuming
+                // diameter large enough and antialiasing over distance
+                // r_outer-r_inner = 1, the difference between simplified:
+                // (d^2-r_inner^2)/(r_outer^2-r_inner^2) and precise: (sqrt(d^2)
+                // - r_inner)/(r_outer-r_inner) is negligible
+
+                if show_all || included {
+                    //active part
+                    if d <= IN_INNER_ANTI {
+                        pix_c_idx = 0;
+                    } else if d <= INNER_MIN {
+                        let c_i = ((15 * (d - IN_INNER_ANTI)) / (INNER_MIN - IN_INNER_ANTI)) as u8;
+                        pix_c_idx = c_i;
+                    } else if d <= OUTER_OUT_ANTI {
+                        pix_c_idx = 15;
+                    } else if d <= OUTER_MAX {
+                        let c_i =
+                            ((15 * (d - OUTER_OUT_ANTI)) / (OUTER_MAX - OUTER_OUT_ANTI)) as u8;
+                        pix_c_idx = 15 - c_i;
+                    } else {
+                        pix_c_idx = 0;
+                    }
+                } else {
+                    //inactive part
+                    if d <= IN_INNER_ANTI {
+                        pix_c_idx = 0;
+                    } else if d <= INNER_MIN {
+                        let c_i = ((15 * (d - IN_INNER_ANTI)) / (INNER_MIN - IN_INNER_ANTI)) as u8;
+                        pix_c_idx = c_i;
+                    } else if d <= INNER_MAX {
+                        pix_c_idx = 15;
+                    } else if d <= INNER_OUTER_ANTI {
+                        let c_i = ((10 * (d - INNER_MAX)) / (INNER_OUTER_ANTI - INNER_MAX)) as u8;
+                        pix_c_idx = 15 - c_i;
+                    } else if d <= OUTER_OUT_ANTI {
+                        pix_c_idx = 5;
+                    } else if d <= OUTER_MAX {
+                        let c_i = ((5 * (d - OUTER_OUT_ANTI)) / (OUTER_MAX - OUTER_OUT_ANTI)) as u8;
+                        pix_c_idx = 5 - c_i;
+                    } else {
+                        pix_c_idx = 0;
+                    }
+                }
+            } else {
+                pix_c_idx = 0;
+            }
+
+            let x = x_c - r.x0;
+            if x % 2 == 0 {
+                pix_c_idx_prev = pix_c_idx;
+            } else {
+                loader_buffer[(x >> 1) as usize] = pix_c_idx_prev | pix_c_idx << 4;
+            }
+        }
+
+        dma2d_wait_for_transfer();
+        dma2d_start_blend(icon_buffer, loader_buffer, r.width());
+    }
+
+    dma2d_wait_for_transfer();
+}
+
+fn process_4bpp_buffer(
```

Comment:
nice, i was actually trying to do this before but didn't think of using const generic and couldn't make it work. 

unfortunately i needed to use two generic params, because doing something like this
`let mut uncomp_buffer = [0u8;  {(constant::WIDTH * BUFFER_BPP) / 8}];`
requires ` #![feature(generic_const_exprs)]` from nightly, which i didn't want to introduce.

https://github.com/trezor/trezor-firmware/commit/c1af69d0561f8b161b09d62d97558fd524ed0d76

Reply:
Never mind, I forgot to add `UI2=1`, apologies

Reply:
agree that we can waste a bit of stack space here and rather save some flash:
https://github.com/trezor/trezor-firmware/commit/a6bef6911a3c75209ff2353fdb0f54c1bf4a7c20